### PR TITLE
feature(rjy): add rough framework of obs encoder

### PR DESCRIPTION
**Pull request by nighood**

### CODE REVIEW

File: genshin_card_env.py [hunk 1]
```diff
@@ -62,6 +64,11 @@ def _get_obs(self, raw_game_info) -> tnp.ndarray:
     obs = env.observation_space.sample()
     print(obs)
     # print([v.shape for v in obs.values()])
+    obs_encoder = ObservationEncoder(env.observation_space)
+    batch_obs = [obs.tensor() for i in range(8)]
+    last_action = env.action_space.sample(obs=obs)
+    batch_last_action = [last_action.tensor() for i in range(8)]
+    encoded_obs = obs_encoder(batch_obs, batch_last_action)
```
Comment (PaParaZz1) [inline genshin_card_env.py]: assert something in `encoded_obs`

File: obs_encoder.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        # global obs
+        self.global_encoder = GlobalObsEncoder(input_size=6)
+        # sub-obs
+        self.dice_encoder = DiceObsEncoder(obs_space)
+        self.character_encoder = CharacterObsEncoder(obs_space)
+        self.skill_encoder = SkillObsEncoder(obs_space)
+        self.card_encoder = CardObsEncoder(obs_space)
+        self.summoner_encoder = SummonerObsEncoder(obs_space)
+        self.supporter_encoder = SupporterObsEncoder(obs_space)
+        self.init_merge_flag = False
```
Comment (PaParaZz1) [inline obs_encoder.py]: why name it to `init_merge_flag`

File: obs_encoder.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+            self.supporter_encoder(new_obs['supporter_obs']),
+            new_obs['supporter_other_info_obs']
+        ], dim=1)
+
+        # merge all obs
+        if not self.init_merge_flag:
+            obs_input_sizes = {}
+            for key,value in encoded_obs.items():
+                obs_input_sizes[key] = value.shape[-1]
+            self.obs_merge = VectorMerge(input_sizes=obs_input_sizes, output_size=256)
```
Comment (PaParaZz1) [inline obs_encoder.py]: define the instance of `nn.Module` in `__init__`

File: obs_encoder.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
+import treetensor.torch as ttorch
```
Comment (PaParaZz1) [inline obs_encoder.py]: add unittest for file, especially taking loss back propogration into consideration

File: obs_encoder.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        supporter_obs_list = []
+        character_other_info_list = []
+        skill_other_info_list = []
+        card_other_info_list = []
+        summoner_other_info_list = []
+        supporter_other_info_list = []
+        for obs, last_action in zip(obs_list, last_action_list):
+            obs = obs.to(torch.float32)
+            last_action = last_action.to(torch.float32)
+            global_obs = torch.cat([
```
Comment (PaParaZz1) [inline obs_encoder.py:218]: use one-hot encoding, then concatanate

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: genshin_card_env.py [hunk 6]
```diff
@@ -2,8 +2,10 @@
 import gym
 import numpy as np
 import treetensor.numpy as tnp
+from ding.torch_utils import to_device
 from obs import get_observation_space, Character, Card
 from action import get_action_space, BasicRuleUtilities
+from obs_encoder import ObservationEncoder
```
Comment (puyuan1996) [inline genshin_card_env.py:8]: 优化import顺序

File: genshin_card_env.py [hunk 7]
```diff
@@ -13,7 +15,7 @@ def __init__(
         env_id: str,
         character_list: Optional[List[Character]],
         card_list: Optional[List[Card]],
-        max_card_num: int = 30,
+        max_card_num: int = 30,     # The deck contains 30 action cards
```
Comment (puyuan1996) [inline genshin_card_env.py]: yapf format

File: genshin_card_env.py [hunk 8]
```diff
@@ -54,17 +56,25 @@ def is_done(self):
...
     for _ in range(5):
         action = env.action_space.sample(obs=obs)
         print(action)
+    batch_size = 8
+    embedding_size = 256
+    obs_encoder = to_device(ObservationEncoder(env.observation_space,output_size=embedding_size),'cuda')
+    batch_obs = to_device([obs.tensor() for i in range(batch_size)],'cuda')
+    batch_last_action = to_device([env.action_space.sample(obs=obs).tensor() for i in range(batch_size)],'cuda')
+    encoded_obs = obs_encoder(batch_obs, batch_last_action)
+    assert encoded_obs.shape == (batch_size, embedding_size), 'shape of encoded_obs wrong'
```
Comment (puyuan1996) [inline genshin_card_env.py]: 'shape of encoded_obs wrong' -> 完整的句子

File: genshin_card_env.py [hunk 9]
```diff
@@ -54,17 +56,25 @@ def is_done(self):
...
+    print([v.shape for v in obs.values()])
     flatten_obs = tnp.concatenate(list(obs.values()))
     print(flatten_obs.shape)
     for _ in range(5):
         action = env.action_space.sample(obs=obs)
         print(action)
+    batch_size = 8
+    embedding_size = 256
+    obs_encoder = to_device(ObservationEncoder(env.observation_space,output_size=embedding_size),'cuda')
+    batch_obs = to_device([obs.tensor() for i in range(batch_size)],'cuda')
```
Comment (puyuan1996) [inline genshin_card_env.py]: batch_obs 由不同的 sample 组成

File: obs_encoder.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
+import treetensor.torch as ttorch
+import torch
+import torch.nn as nn
+from ding.torch_utils import to_device
+from obs import ObservationSpace
+from ding.torch_utils.network import VectorMerge
+
+class SubObsEncoder(nn.Module):
+    def __init__(self, input_size, output_size=256, hidden_size=64):
```
Comment (puyuan1996) [inline obs_encoder.py]: 添加type hint

Comment (puyuan1996) [reply obs_encoder.py]: 为每个类添加overview comments

File: obs_encoder.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    def forward(self, gloabl_obs):
+        return self.encoder(gloabl_obs)
+
+class ObservationEncoder(nn.Module):
+    def __init__(self, obs_space:ObservationSpace, output_size=256, hidden_size=256):
+        super(ObservationEncoder, self).__init__()
+
+        # global obs
+        self.global_encoder = GlobalObsEncoder(input_size=56)
+        # sub-obs
```
Comment (puyuan1996) [inline obs_encoder.py:145]: 更加详细的注释

File: obs_encoder.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+        character_obs_list = []
+        skill_obs_list = []
+        card_obs_list = []
+        summoner_obs_list = []
+        supporter_obs_list = []
+        character_other_info_list = []
+        skill_other_info_list = []
+        card_other_info_list = []
+        summoner_other_info_list = []
+        supporter_other_info_list = []
```
Comment (puyuan1996) [inline obs_encoder.py]: 或许可以换成
```
processed_obs_lists = {
    "global": [],
    "dice": [],
    "character": [],
    "skill": [],
    "card": [],
    "summoner": [],
    "supporter": [],
    "character_other_info": [],
    "skill_other_info": [],
    "card_other_info": [],
    "summoner_other_info": [],
    "supporter_other_info": [],
}
```

File: obs_encoder.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+        skill_obs = torch.stack(skill_obs_list, dim=0)
+        card_obs = torch.stack(card_obs_list, dim=0)
+        summoner_obs = torch.stack(summoner_obs_list, dim=0)
+        supporter_obs = torch.stack(supporter_obs_list, dim=0)
+
+        character_other_info_obs = torch.stack(character_other_info_list, dim=0)
+        skill_other_info_obs = torch.stack(skill_other_info_list, dim=0)
+        card_other_info_obs = torch.stack(card_other_info_list, dim=0)
+        summoner_other_info_obs = torch.stack(summoner_other_info_list, dim=0)
+        supporter_other_info_obs = torch.stack(supporter_other_info_list, dim=0)
```
Comment (puyuan1996) [inline obs_encoder.py]: 在使用dict存储信息后，这里可以修改为
```
processed_obs_stacks = {}
for key in processed_obs_lists.keys():
    processed_obs_stacks[key] = torch.stack(processed_obs_lists[key], dim=0)
```

File: test_obs_encoder.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        # obs = env.observation_space.sample()
+        # print(obs)
+        obs_encoder = to_device(ObservationEncoder(env.observation_space, output_size=embedding_size), test_device)
+        batch_obs = [env.observation_space.sample() for i in range(B)]
+        batch_last_action = to_device([env.action_space.sample(obs=obs).tensor() for obs in batch_obs], test_device)
+        batch_obs_tensor = to_device([obs.tensor() for obs in batch_obs], test_device)
+        for obs,last_action in zip(batch_obs_tensor, batch_last_action):
+            print(obs, last_action)
+        encoded_obs = obs_encoder(batch_obs_tensor, batch_last_action)
+        assert encoded_obs.shape == (B, embedding_size), 'shape of encoded_obs wrong'
```
Comment (puyuan1996) [inline test_obs_encoder.py]: assert信息写完整一些，并给出报错后给用户的排查方案

File: obs_encoder.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+            new_obs['card_other_info_obs']
+        ], dim=1)
+        encoded_obs['summoner_obs'] = torch.cat([
+            self.summoner_encoder(new_obs['summoner_obs']),
+            new_obs['summoner_other_info_obs']
+        ], dim=1)
+        encoded_obs['supporter_obs'] = torch.cat([
+            self.supporter_encoder(new_obs['supporter_obs']),
+            new_obs['supporter_other_info_obs']
+        ], dim=1)
```
Comment (puyuan1996) [inline obs_encoder.py]:
```
# 先在 __init__中定义：

self.encoders = {
    'global': self.global_encoder,
    'dice': self.dice_encoder,
    'character': self.character_encoder,
    'skill': self.skill_encoder,
    'card': self.card_encoder,
    'summoner': self.summoner_encoder,
    'supporter': self.supporter_encoder
}

# 然后将上面的代码修改为：

# Handle global and dice separately as they don't need concatenation
encoded_obs['global_obs'] = self.encoders['global'](new_obs['global_obs'])
encoded_obs['dice_obs'] = self.encoders['dice'](new_obs['dice_obs'])

# Loop over the rest of the keys
for key in ['character', 'skill', 'card', 'summoner', 'supporter']:
    encoded_obs[f'{key}_obs'] = torch.cat([
        self.encoders[key](new_obs[f'{key}_obs']),
        new_obs[f'{key}_other_info_obs']
    ], dim=1)
```

File: obs_encoder.py (hunk 11 above)
Comment (PaParaZz1) [reply obs_encoder.py:145]: 全面的注释先不加了哈，因为可能还要修改，只加一写关键的注释就行

File: obs_encoder.py [hunk 16]
```diff
@@ -19,7 +25,11 @@ def forward(self, sub_obs):
         return self.sub_encoder(sub_obs)
 
 class DiceObsEncoder(nn.Module):
-    def __init__(self, obs_space, output_size=256):
+    r"""
+    Overview:
+        Encoder that encodes dice_obs.
```
Comment (puyuan1996) [inline obs_encoder.py:30]: 每个overview里面关于obs的类别或许可以解释再详细一点，达到对于不熟悉七圣召唤游戏的人也能大概知道其含义。

File: obs_encoder.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+        obs_input_sizes = {
+            'global_obs': self.global_encoder.output_size,
+            'dice_obs': self.dice_encoder.output_size,
+            'character_obs': self.character_encoder.output_size+obs_space.character_other_info.shape[0],
+            'skill_obs': self.skill_encoder.output_size+obs_space.skill_other_info.shape[0],
+            'card_obs': self.card_encoder.output_size+obs_space.card_other_info.shape[0],
+            'summoner_obs': self.summoner_encoder.output_size+obs_space.summoner_other_info.shape[0],
+            'supporter_obs': self.supporter_encoder.output_size+obs_space.supporter_other_info.shape[0],
+        }
+        self.obs_merge = VectorMerge(input_sizes=obs_input_sizes, output_size=output_size)
```
Comment (puyuan1996) [inline obs_encoder.py:171]: 增加注释
